common/achievement/evaluate.ts

```
@@ -0,0 +1,114 @@
+        }
+        // we take the relation from the first event, that posesses one, in order to create buckets from it, if needed
+        const relation = eventsForMetric.find((event) => event.relation)?.relation;
+
+        const bucketCreatorFunction = bucketCreatorDefs[bucketCreator].function;
+        const bucketAggregatorFunction = aggregators[bucketAggregator].function;
+
+        const aggregatorFunction = aggregators[aggregator].function;
+
+        if (!bucketCreatorFunction || !bucketAggregatorFunction || !aggregatorFunction) {
+            logger.error(
+                `No bucket creator or aggregator function found for ${bucketCreator}, ${aggregator} or ${bucketAggregator} during the evaluation of achievement`
+            );
+            return;
+        }
+
+        let bucketContext: AchievementContextType;
+        if (relation) {
+            bucketContext = await getBucketContext(relation);
+        }
+
+        const buckets = bucketCreatorFunction({ recordValue: recordValue, context: bucketContext });
+
+        const bucketEvents = createBucketEvents(eventsForMetric, buckets);
+        const bucketAggr = bucketEvents.map((bucketEvent) => bucketAggregatorFunction(bucketEvent.events.map((event) => event.value)));
+
+        const value = aggregatorFunction(bucketAggr);
+        resultObject[key] = value;
+    }
+    // TODO: return true if the condition is empty (eg. a student finishes a course and automatically receives an achievement)
+    const evaluate = swan.parse(condition);
+    const value: boolean = await evaluate(resultObject);
+
+    return {
+        conditionIsMet: value,
+        resultObject,
+    };
+}
+
+export function createBucketEvents(events: Achievement_event[], bucketConfig: BucketConfig): BucketEvents[] {
+    switch (bucketConfig.bucketKind) {
+        case 'default':
+            return createDefaultBuckets(events, bucketConfig);
+        case 'time':
+            return createTimeBuckets(events, bucketConfig);
+    }
```

please throw an error if none was found.

---

The case could not be nothing, following the code before. But, if it was, we return the default bucket events now.

---

My idea was that it would save us from typo's when adding new buckets in the future.
I just checked, and the typesystem would catch it. Good job 👍